### PR TITLE
RDFPFN792026: preserve guarded createRef persistence

### PR DESCRIPTION
**Pull request by aidenybai**
## Why

`no-create-ref-in-function-component` reported three `createRef()` calls that are initialized once into a persistent React `useRef().current` value. React preserves the outer ref object across renders, and the exact empty-value guard prevents later renders from replacing the nested refs.

React Bench evidence: `fix-react-cloudscape-design-comp__mun7kHG`, `src/app-layout/visual-refresh-toolbar/visual-refresh-toolbar.tsx/before-main-slot.tsx:35`.

- evidence SHA-256: `dd5101abb13ff47b037aa4b715b700162664324e6c22b436fba2f033ff1c6c8e`
- reconstructed source SHA-256: `b07ffb8d4de282140eb407725e51b093e7b29a2c40441d9885f1e41d4609d576`
- exact replay: current main `3` diagnostics → this branch `0`

## What changed

- Prove only direct or object-literal `createRef()` values persisted by the sole matching empty-ref guard.
- Require a resolved React `useRef`, immutable local binding, unconditional hook initialization, matching sentinel, and no later `.current` writes.
- Keep reporting mismatched guards, observed-before-persistence values, nested conditional writes, mutable or shadowed refs, conditional hooks, and later clears or replacements.
- Freeze the exact React Bench source and hashes in regression coverage.
- Add a strict-fuzz regression seed and a patch changeset.

## Validation

- `nr build`
- `nr test`
- `nr typecheck`
- `nr lint`
- `nr format:check`
- `nr smoke:json-report`
- `FUZZ_RULE=no-create-ref-in-function-component FUZZ_STRICT=1 FUZZ_ITERATIONS=500 nr fuzz`
- exact React Bench replay and adversarial positive/negative controls

Local RDE could not yield diagnostic evidence because its current parser accepts schema version 1 while current React Doctor emits schema version 3; every attempted repository was rejected before diagnostics. Daytona parity is intentionally pending behind the existing shared queue.

Tracking: RDFPFN792026

---

> [!NOTE]
> **Low Risk**
> Lint-only heuristic change with broad negative tests; risk is under-reporting unsafe `createRef` in edge cases, not runtime behavior.
> 
> **Overview**
> **`no-create-ref-in-function-component`** no longer warns when `createRef()` is stored once on a stable **`useRef().current`** behind a matching empty-ref guard (e.g. `if (!ref.current) { ref.current = { … createRef() … } }`).
> 
> A new **`isCreateRefPersistedInGuardedUseRef`** helper proves that pattern with scope/CFG checks: real `useRef`, const binding, unconditional hook init, guard aligned with the initial sentinel (`null` / `false` / `undefined`), direct or object-literal persistence, and no later `.current` writes (including via const aliases). Unsafe variants still report.
> 
> Coverage adds a React Bench fixture replay, adversarial regression tests, a fuzz corpus seed, and a patch changeset.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit d62408a7e60c5201fa38e43edff043fb6d28953d. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: packages/oxlint-plugin-react-doctor/src/plugin/rules/react-builtins/is-create-ref-persisted-in-guarded-use-ref.ts [hunk 1]
```diff
@@ -0,0 +1,201 @@
...
+  scopes: ScopeAnalysis,
+): EsTreeNode | null => {
+  const memberExpression = identifier.parent;
+  if (
+    !isNodeOfType(memberExpression, "MemberExpression") ||
+    memberExpression.object !== identifier ||
+    !isCurrentMemberForSymbol(memberExpression, symbol, scopes)
+  ) {
+    return null;
+  }
```
[the remark this answers is not in the document]
Comment (aidenybai) [reply packages/oxlint-plugin-react-doctor/src/plugin/rules/react-builtins/is-create-ref-persisted-in-guarded-use-ref.ts:133]: Fixed in 9bff86ca2. Later-write detection now unwraps transparent receiver expressions before matching `.current`, with regression coverage for parenthesized, TypeScript-asserted, and non-null receivers. The affected plugin suite, typecheck, fuzz corpus, strict 500-iteration fuzz run, lint, and format checks all pass.

File: packages/oxlint-plugin-react-doctor/src/plugin/rules/react-builtins/is-create-ref-persisted-in-guarded-use-ref.ts [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+    refInitializer.arguments.length === 0
+      ? "undefined"
+      : getEmptyRefValueKind(refInitializer.arguments[0], scopes);
+  if (!initialValueKind) return false;
+  const guard = findGuard(assignment);
+  if (!guard || !doesGuardProveEmptyRef(guard, refSymbol, initialValueKind, scopes)) return false;
+  for (const reference of refSymbol.references) {
+    const write = isDirectCurrentWrite(reference.identifier, refSymbol, scopes);
+    if (write && write !== assignment) return false;
+  }
```
[the remark this answers is not in the document]
Comment (aidenybai) [reply packages/oxlint-plugin-react-doctor/src/plugin/rules/react-builtins/is-create-ref-persisted-in-guarded-use-ref.ts:237]: Fixed in d62408a7e. The persistence proof now walks the existing transitive const-alias graph, rejects later `.current` writes through any alias, and conservatively rejects outer-ref escapes it cannot prove safe. Regression coverage includes an alias chain write and an opaque callback escape. The affected plugin suite (963 files / 25,023 tests), typecheck, fuzz corpus, strict 500-iteration fuzz run, lint, format, and post-change deduplication all pass.
